2017/gonumtour.slide

```diff
@@ -0,0 +1,300 @@
+		...
+		Concurrent bool // Should the function calls be executed concurrently.
+	}
+
+* Toward a Solution to the Two Language Problem
+- Pure Go - No mandatory C/Fortran
+- Translated 125+ LAPACK and BLAS functions
+- Competitive in speed (especially for full programs)
+
+* Gonum Design Philosophy
+
+* Panics vs. Errors
+- Rule #1: NO SILENT BUGS - Input is either clearly correct or it is wrong
+
+- Panic: Error is easy to check before call
+
+	// Add adds, element-wise, the elements of s and dst, and stores in dst.
+	// Panics if the lengths of dst and s do not match.
+	func Add(dst, s []float64) {
+		...
+	}
+
+- Error: Impossible to check without performing the operation
+
+	// Solve finds a minimum-norm solution to a system of linear equations defined
+	// by the matrices a and b. If A is singular or near-singular, a Condition error
+	// is returned.
+	func (m *Dense) Solve(a, b Matrix) error {
+		...
+	}
+
+* Memory Re-Use
+- If a function creates a type, allow a destination as a first argument
+- But allow a nil destination for ease
+
+	// Rand generates a random number according to the distributon.
+	// If the input slice is nil, new memory is allocated, otherwise the result is stored
+	// in place.
+	func (n *Normal) Rand(x []float64) []float64 {
+
+   
+	// CovarianceMatrix returns the covariance matrix calculated from a matrix of data, x.
+	//
+	// If cov is not nil it must either be zero-sized or have the same number of
+	// columns as the input data matrix. cov will be used as the destination for
+	// the covariance data. If cov is nil, a new mat.SymDense is allocated for
+	// the destination.
+	func CovarianceMatrix(cov *mat.SymDense, x mat.Matrix, weights []float64) *mat.SymDense {
+
+* "Functional" Programming
+- Functions and methods should not modify state, unless that is their only purpose
+- Even at the cost of extra allocations
+- Really important for easy concurrent programming
+
+* The Gonum Suite Functionality
+
+* Repositories
+`gonum.org/v1/gonum`: Main repository
+
+`github.com/gonum/plot`: Plotting library (will change to gonum.org import in the future)
+
+A few other repositories for `cgo` wrappers 
+
+- `gonum.org/v1/netlib`: LAPACK and BLAS wrappers (not mandatory)
+- `gonum.org/v1/hdf5`: For using the [[https://support.hdfgroup.org/HDF5/][hdf5]] file format
+
+* Package Overview
+
+- `floats`: Operations on `float64` and (pairs of) `[]float64` (Add, Norm, Distance)
+- `mat`: Matrix interface and types (Dense, SymDense) + Linear Algebra (Solve, QR)
+- `stat` (and subpackages): Stastics functions, probability distributions, sampling routines, spatial correlations
+- `graph`: Constructing, analyzing and manipulating (network) graphs
+- `optimize`: Finding the minima of functions
+- `diff`/`integrate`: Finding the numerical derivative/integral of functions
+
+* floats
+
+* floats (http://godoc.org/gonum.org/v1/gonum/floats)
+- Straightforward functionality, browse the functions
+
+.play gonumtour/floats.go
```

Comment:
I'm guessing this is one of the places you think I can clean up the dead lines of code. I had a couple comments from people at the event that they liked seeing the full import statement, and it also highlights the updated import locations. I'd like to keep the header here.